### PR TITLE
Add support for GlassFish 7.0.6, 7.0.7, and 7.0.9

### PR DESCRIPTION
**Pull request by pepness**
Library Notes:

- Add support for Java 21
- Refactoring and maintenance of code base
- Fix processing of the Multi-Release libraries in EAR application
- Bug fixes and updates of various components

NetBeans GlassFish module notes:

- Add support for GlassFish version 7.0.7, 7.0.8, and 7.0.9
- Add new ConfigBuilderProvider for GlassFish 7.0.9 and newer versions that support Java 21
- Add support for JavaSE constants up to version 22
- Bump `javac.source` and `javac.target` to version 11 for all 4 GlassFish modules `common`, `eecommon`, `javaee`, and `tooling`
- Refactor creation of `ServerWizardIterator` with one private array
- Remove some unnecessary imports
- Make final some fields
- Refactor long switch statement to one `if if-else else` statement
- Use ConcurrentMap for `builders`, set initial capacity, use atomic operations `computeIfPresent` and `computeIfAbsent`, and remove some synchronized blocks because the ConcurrentHashMap implementation already use synchronized at a field level

NetBeans Testing:

- Verify successful execution of libraries and licenses Ant test
- Verify successful execution of Verify Sigtests
- Verify successful execution of `ant -Dcluster.config=release commit-validation`
- Verify successful execution of unit tests for modules `glassfish.common`, `glassfish.javaee`, `glassfish.tooling`, and `glassfish.eecommon`
- Started NetBeans and ensure the log didn't have any ERROR or new WARNINGS
- Successfully register GlassFish 7.0.7, 7.0.8, and 7.0.9, create a web app and verify that it works.

[Release Notes](https://github.com/eclipse-ee4j/glassfish/releases/tag/7.0.9)

### CODE REVIEW

File: enterprise/glassfish.common/src/org/netbeans/modules/glassfish/common/ServerDetails.java [hunk 1]
```diff
@@ -480,6 +459,9 @@ public static int getVersionFromInstallDirectory(File glassfishDir)  {
                 case GF_7_0_4:   return GLASSFISH_SERVER_7_0_4.getVersion();
                 case GF_7_0_5:   return GLASSFISH_SERVER_7_0_5.getVersion();
                 case GF_7_0_6:   return GLASSFISH_SERVER_7_0_6.getVersion();
+                case GF_7_0_7:   return GLASSFISH_SERVER_7_0_7.getVersion();
+                case GF_7_0_8:   return GLASSFISH_SERVER_7_0_8.getVersion();
+                case GF_7_0_9:   return GLASSFISH_SERVER_7_0_9.getVersion();
```
Comment (mbien) [inline enterprise/glassfish.common/src/org/netbeans/modules/glassfish/common/ServerDetails.java:464]: can this switch be eliminated by putting `GLASSFISH_SERVER_` enums into `GF_7_` enums during construction?

File: enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java [hunk 2]
```diff
@@ -193,16 +209,13 @@ public static ConfigBuilder getBuilder(final GlassFishServer server) {
...
-            builder = new ConfigBuilder(config, serverHome, serverHome, serverHome);
-            builders.put(server, builder);
+        builder = builders.computeIfPresent(server, (key, value) -> value);
+        if (builder != null) {
+            return builder;
         }
-        return builder;
+        String serverHome = server.getServerHome();
+        return builders.computeIfAbsent(server, key -> 
+                new ConfigBuilder(config, serverHome, serverHome, serverHome));
```
Comment (mbien) [inline enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java:218]: something is wrong here. Please take another look.
`computeIfPresent(server, (key, value) -> value)` is essentially `get()` and I believe this was meant to be removed?

Comment (pepness) [reply enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java:218]: The `computeIfPresent` implementation from `ConcurrentHashMap` synchronize at a node level (`Node<K,V>`).  I use `computeIfPresent` to circumvent the need of synchronize the entire map. It does work like a `get()` (`get()` does not synchronize) and it does not have the `@Deprecated` as of now.

Comment (mbien) [reply enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java:218]: the old version:
```java
synchronized {
    // checked if `server` was in the map, if it was, it returned it.
    // if it wasn't in the map, it created a new ConfigBuilder, registered it in the map and returned it
}
```
this is the same as:
```java
        String serverHome = server.getServerHome();
        return builders.computeIfAbsent(server, key -> 
                new ConfigBuilder(config, serverHome, serverHome, serverHome));
```
(If i am missing something, please explain)

Comment (pepness) [reply enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java:218]: Both versions will compute the same results. But the old version does not allow concurrent modifications to the map, the new version will allow concurrent modifications to the map.
Old version:
```java
synchronized (entireMap) {
    // checked if `server` key was in the map, if it was, it returned it's value.
    // if it wasn't in the map, it created a new ConfigBuilder, registered it in the map and returned it
}
```
New version restrict modifications at a node level e.g. `Node[K1,V1]` until it finish, but allows  concurrent modifications to every other node e.g. `Node[K2,V2]`:
```java
synchronized (Node[K1,V1]) { 
    // Thread 1
    // No one can modify this node until I finish
}
synchronized (Node[K2,V2]) {
    // Thread 2
    // Modifying this node at the same time that Thread 1 is working on Node[K1,V1]
}
```

Comment (mbien) [reply enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java:218]: the the task of the method, which is to return the value of the map or compute it if it isn't there yet, can be written with one operation on the map:
```java
    public static ConfigBuilder getBuilder(final GlassFishServer server) {
        if (server == null) {
            throw new ServerConfigException(
                    "GlassFish server entity shall not be null");
        }
        String serverHome = server.getServerHome();
        return builders.computeIfAbsent(server, key -> 
                new ConfigBuilder(config, serverHome, serverHome, serverHome));
    }

```

Comment (pepness) [reply enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java:218]: You are right!, everything will be done in one atomic and concurrent statement, great work @mbien.
I will re-test everything and push the update.

Comment (mbien) [reply enterprise/glassfish.tooling/src/org/netbeans/modules/glassfish/tooling/server/config/ConfigBuilderProvider.java:218]: figured out how to create a jackpot rule for this, nb can refactor it now with one click:
https://github.com/mbien/jackpot-inspections/commit/8eadafc5c45c3a1898ebadaa16455c5019537c9c

File: enterprise/glassfish.common/nbproject/org-netbeans-modules-glassfish-common.sig [hunk 3]
```diff
@@ -620,6 +635,8 @@ CLSS public abstract interface java.util.EventListener
 
 CLSS public abstract interface java.util.Map<%0 extends java.lang.Object, %1 extends java.lang.Object>
 innr public abstract interface static Entry
+meth public !varargs static <%0 extends java.lang.Object, %1 extends java.lang.Object> java.util.Map<{%%0},{%%1}> ofEntries(java.util.Map$Entry<? extends {%%0},? extends {%%1}>[])
+ anno 0 java.lang.SafeVarargs()
 meth public abstract boolean containsKey(java.lang.Object)
```
Comment (mbien) [inline enterprise/glassfish.common/nbproject/org-netbeans-modules-glassfish-common.sig:640]: @ebarboni @matthiasblaesing why are there JDK classes in the sig file of this module?

Comment (mbien) [reply enterprise/glassfish.common/nbproject/org-netbeans-modules-glassfish-common.sig:640]: the sig generator is probably not filtering out JDK classes, which means if the public API uses JDK classes in declarations, the sig changes when JDK level changes.

**Comment by mbien** [comment]
@pepness can you squash so that it is ready for merge?

**Review by mbien — APPROVED** [review]
good cleanup! Should make future updates easier I think.

**Pull request merged**